text/0000-layout-packed-aligned.md

```diff
@@ -0,0 +1,182 @@
+- Feature Name: `layout_packed_aligned`
+- Start Date: 2024-10-24
+- RFC PR: [rust-lang/rfcs#3718](https://github.com/rust-lang/rfcs/pull/3718)
+- Rust Issue: [rust-lang/rust#100743](https://github.com/rust-lang/rust/issues/100743)
+
+# Summary
+[summary]: #summary
+
+This RFC makes it legal to have `#[repr(C)]` structs that are:
+- Both packed and aligned.
+- Packed, and transitively contains`#[repr(align)]` types.
+
+It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.
+It has the same behavior as `#[repr(C)]` except on `*-pc-windows-gnu` targets where it uses the msvc layout
+rules instead.
+
+# Motivation
+[motivation]: #motivation
+
+This RFC enables the following struct definitions:
+
+```rs
+#[repr(C, packed(2), align(4))]
+struct Foo { // Alignment = 4, Size = 8
+    a: u8,   // Offset = 0
+    b: u32,  // Offset = 2
+}
+```
+
+This is commonly needed when Rust is being used to interop with existing C and C++ code bases, which may contain
+unaligned types. For example in `clang` it is possible to create the following type definition, and there is
+currently no easy way to create a matching Rust type:
+
+```cpp
+struct  __attribute__((packed, aligned(4))) MyStruct {
+    uint8_t a;
+    uint32_t b;
+};
+```
+
+Currently, `#[repr(packed(_))]` structs cannot be `#[repr(align(_))]` or transitively contain `#[repr(align(_))]` types. Attempting to do so results in a [hard error](https://doc.rust-lang.org/nightly/error_codes/E0588.html).
+
+This behavior was added in the [original implementation](https://github.com/rust-lang/rust/issues/33158) of `#[repr(packed)]` due to concerns over differing behavior between msvc and gcc/clang. This makes it cumbersome or even impossible to produce C-compatible struct layouts in Rust when the corresponding C types were annotated with both `packed` and `aligned`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## `#[repr(C)]`
+When `align` and `packed` attributes exist on the same type, or when `packed` structs transitively contains `align` types,
+the resulting layout matches the target toolchain ABI.
+
+For example, given:
+```c
+#[repr(C, align(4))]
+struct Foo(u8);
+#[repr(C, packed(1))]
+struct Bar(Foo);
+```
+`align_of::<Bar>()` would be 4 for `*-pc-windows-msvc` and 1 for everything else.
```

Comment:
IMO this is extremely unintuitive behavior, especially for those of us writing custom derives (like [zerocopy](https://crates.io/crates/zerocopy)). We can't just rely on querying `align_of` - we need to be able to parse a type's definition and understand the semantics of any `repr` attributes on it.

Also, while this is technically not a breaking change, I wouldn't be surprised if some custom derive code implicitly assumes that this behavior isn't possible, and would become unsound in the face of this change. For example, as it stands today, it is valid to assume that a `#[repr(align(N))]` type has alignment at least `N`, but that would stop being valid with this change.

I'd propose that if system-specific behavior like this is required, it'd be better to do it behind a new repr so that the behavior of `repr(C)` remains straightforward.

Reply:
> as it stands today, it is valid to assume that a `#[repr(align(N))]` type has alignment at least `N`, but that would stop being valid with this change

Did you say this wrong somehow? By this RFC, `align(N)` types are still always aligned to at least `N` when behind a reference. In fact, they're aligned to `N` in more cases, as `packed` types can't underalign them with MSVC layout rules.

What changes is that `packed(M)` types no longer have an alignment of exactly `M`, but instead have an alignment of at least `M`.

> custom derive code might be unsound

Such custom derive code is already likely quite iffy, since `#[derive(Trait)] #[proc_macro] struct` will have the derive see the token stream before the proc macro processes it and can modify all the decorated code arbitrarily. `zerocopy` is aware of this and doesn't permit any attribute it doesn't know the reserved semantics of to be in the type definition, but that's a high, difficult, and restrictive bar to clear most of the time.

Reply:
> > as it stands today, it is valid to assume that a `#[repr(align(N))]` type has alignment at least `N`, but that would stop being valid with this change
> 
> Did you say this wrong somehow? By this RFC, `align(N)` types are still always aligned to at least `N` when behind a reference. In fact, they're aligned to `N` in more cases, as `packed` types can't underalign them with MSVC layout rules.
> 
> What changes is that `packed(M)` types no longer have an alignment of exactly `M`, but instead have an alignment of at least `M`.

You're right, I misread the text. Still, there's a concern: Our derives currently take `#[repr(C, packed(N))]` as a guarantee that the decorated type cannot have alignment greater than `N`. This is consistent with [the Reference](https://doc.rust-lang.org/reference/type-layout.html#the-alignment-modifiers):

> For `packed`, ... the alignments of each field, for the purpose of positioning fields, is the smaller of the specified alignment and the alignment of the field’s type.

IIUC, to make `packed` no longer provide this guarantee is a breaking change with respect to the Reference.

---

There's another concern as well: This makes "whether or not the type has an alignment repr (`packed`/`align`)" part of the type's layout. The Reference currently specifies how types are laid out purely in terms of the sizes and alignments of their field types, and not in terms of any other facts about the field types. With this change, the following two `Bar` types would have different layouts despite containing `Foo` types with identical sizes and alignments:

```rust
#[repr(C, align(4))]
struct Foo(u8);
#[repr(C, packed(1))]
struct Bar(Foo);

#[repr(C)]
struct Foo(u8, [u32; 0]);
#[repr(C, packed(1))]
struct Bar(Foo);
```

Reply:
> ```rust
> #[repr(C, align(4))]
> struct Foo(u8);
> #[repr(C, packed(1))]
> struct Bar(Foo);
> ```

This does not compile today (reference said "a `packed` type cannot transitively contain another `aligned` type"), so we are free to choose whether the two `Bar`s should have same or different layouts after this RFC.

```
error[E0588]: packed type cannot transitively contain a `#[repr(align)]` type
 --> src/lib.rs:4:1
  |
4 | struct Bar(Foo);
  | ^^^^^^^^^^
  |
note: `Foo` has a `#[repr(align)]` attribute
 --> src/lib.rs:2:1
  |
2 | struct Foo(u8);
  | ^^^^^^^^^^
```

Reply:
My point isn't that this is a breaking change, but rather than it breaks the current *mental model*, which is that a type's layout is determined solely by the size and alignment of its fields. This RFC changes that to say that a type's layout is determined by the size, alignment, *and representation* of its fields. Since representation is a fairly complicated concept (it can include repr(C), repr(Int), repr(packed), repr(align), repr(transparent), and some-but-not-all combinations of these), IMO it significantly complicates the mental model of type layout to say that a type's layout also depends upon its fields' representations.

Reply:
Unfortunately, the existing. mental model turned out to be in contradiction to reality, at least on MSVC targets. When you build a model and then reality demonstrates the incorrectness of the model, sometimes the best option is to fix your model. Yes we documented this model in a bunch of places, but... what else could we do? Stick our head into the sand and continue pretending that the simpler model is "good enough"?

So now the layout of a type is determined by the size, alignment, and [*explicitly requested alignment*](https://github.com/rust-lang/rfcs/pull/3718#discussion_r1826095670) of its fields.

Reply:
> Unfortunately, the existing mental model turned out to be in contradiction to reality, at least on MSVC targets. When you build a model and then reality demonstrates the incorrectness of the model, sometimes the best option is to fix your model.

There is another alternative: argue that because `__declspec(align)` and `#pragma pack` are nonstandard C extensions, we don't necessarily need to be layout-compatible with the system C in these cases. That's basically the argument for a struct with one zero-length array member on MSVC being zero-sized and not align-sized.

Counter to my own point, though, is that `struct alignas(N) Name` *is* standard C*++*, and the MSVC behavior is the same as with `declspec` in C mode. `#pragma pack` is still nonstandard (and I find MSVC's definition of it to be bonkers) but standard type alignas significantly weakens the argument that it isn't necessary to support. 

But if there exists a type on MSVC with basic alignment higher than the default member packing level without also having an explicitly requested alignment (thus not being sufficiently aligned by default), then I would swing towards writing of MSVC layout edge cases as broken.

> This does not compile today

Unfortunately,​ this error is [straightforward to sidestep with generics.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d713f3ab07be7acd453da00eff0d18a7)

```rust
#[repr(align(4))]
struct Aligned;

#[repr(packed(1))]
struct Packed<T = Aligned>(T);

fn main() {
    dbg!(align_of::<Packed>());
    // [src/main.rs:8:5] align_of::<Packed>() = 1
}
```

Reply:
Isn't it the case that all of these issues could be avoided by introducing a new repr rather than changing the meaning of `repr(C)`? I understand that, historically, `repr(C)` was intended to mean "does the same thing as the C compiler", but it's since come to be used for *much* more beyond that, and its current semantics are relied upon by a huge amount of `unsafe` code that would become silently unsound in the face of the changes proposed here. Is there a reason that introducing a new repr (rather than changing the meaning of `repr(C)`) wouldn't be the obvious solution to address all of the present concerns?

Reply:
The counter argument is that a similarly huge amount of `unsafe` code exists which is relying on `repr(C)` meaning compatible with the target C compiler's structure layout. It comes down to which use case you consider more important.

This RFC's position is that `repr(C)` matching C is the more important property, and I posit that there's more generated C bindings that wouldn't get migrated to a new repr than there is relying on the exact current behavior; the latter usage is much more likely to be actively maintained and tracking this kind of development, whereas generated bindings are perceived more stable.

My personal position is that `repr(C)` matching the target C compiler is only relevant limited to standard C compliant structure layout, but I do see the benefits of both sides of the debate. It's unfortunately similar to splitting size_t/uintptr_t in that Rust assumes the two usages are fully identical, but edge cases mean that that aren't.